src/lando/main/scm/git.py

```diff
@@ -0,0 +1,298 @@
+
+        if self.is_initialized:
+            raise
+
+        result = self._git_run("clone", self.pull_path, self.name)
+        if result.returncode == 0:
+            self.is_initialized = True
+            self.save()
+        else:
+            raise Exception(f"{result.returncode}: {result.stderr}")
+
+    def _git_pull(self):
+        self._git_run("pull", "--all", "--prune")
+
+    def _git_apply_patch(
+        self, diff: str, commit_description: str, commit_author: str, commit_date: str
+    ):
+        f_msg = tempfile.NamedTemporaryFile(encoding="utf-8", mode="w+")
+        f_diff = tempfile.NamedTemporaryFile(encoding="utf-8", mode="w+")
+        with f_msg, f_diff:
+            f_msg.write(commit_description)
+            f_msg.flush()
+            f_diff.write(diff)
+            f_diff.flush()
+
+            self._git_run("apply", f_diff.name)
+
+            self._git_run("add", "-A")
+            self._git_run(
+                "commit",
+                "--date",
+                commit_date,
+                "--author",
+                commit_author,
+                "--file",
+                f_msg.name,
+            )
+
+    def _git_last_commit_id(self) -> str:
+        return self._git_run("rev-parse", "HEAD", cwd=self.path)
+
+    def _git_run(self, *args, cwd: Optional[str] = None, must_succeed: bool = True):
+        """Run a git command and return full output.
+
+        Parameters:
+
+        args: list[str]
+            Arguments to git
+
+        cwd: str
+            Optional path to work in, default to self.path
+
+        Returns: CompletedProcess[str]
+
+            An object with (at least) the following attributes: args, returncode, stdout and stderr
+        """
+        path = cwd or self.path
+        command = ["git"] + list(args)
+        logger.debug("Running " + " ".join(command) + " in " + path)
+        result = subprocess.run(
+            command, cwd=path, capture_output=True, text=True, env=self._git_env()
+        )
+
+        if must_succeed and result.returncode:
+            raise SCMException(
+                f"Error running git command; {command=}, {path=}, {result.stderr}",
+                result.stdout,
+                result.stderr,
+            )
+        return result.stdout.strip()
+
+    @staticmethod
+    def _git_call(*args, cwd: str) -> bool:
+        """Run a git command and return a boolean indicating success or failure.
+
+        Parameters:
+
+        args: list[str]
+            Arguments to git
+
+        cwd: str
+            NON OPTIONAL path to work in, default to self.path
+
+        WARNING
+
+        Returns: boolean
+
+            A success indicator
+
+        """
+        command = ["git"] + list(args)
+        logger.debug("Calling " + " ".join(command) + " in " + cwd)
+        returncode = subprocess.call(
+            command,
+            cwd=cwd,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
```

Comment:
Why suppress the output? Might be useful to keep some of this in the logs, at least for debugging.

Reply:
This method is only used to get the return code for early `git`-based checks (it's also a `@staticmethod` so it can be called when we don't have an instance yet).

We can capture the output, but I'm not sure how we would provide it back. We don't throw exception for non-0 return code, so we don't have something to transport the output to the caller.

Reply:
Ok, I found a better way! https://github.com/mozilla-conduit/lando/pull/147#discussion_r1872351301